File: api/internal/repository/workspace_repository.go
package repository

import (
	"context"

	workspacesv1 "github.com/kloudlite/kloudlite/api/internal/controllers/workspace/v1"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

// WorkspaceRepository provides operations for Workspace resources (namespace-scoped)
type WorkspaceRepository interface {
	NamespacedRepository[*workspacesv1.Workspace, *workspacesv1.WorkspaceList]

	// Domain-specific methods
	GetByOwner(ctx context.Context, owner string, namespace string) (*workspacesv1.WorkspaceList, error)
	GetByWorkMachine(ctx context.Context, workMachineName string, namespace string) (*workspacesv1.WorkspaceList, error)
	ListActive(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error)
	ListSuspended(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error)
	ListArchived(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error)
	SuspendWorkspace(ctx context.Context, name string, namespace string) error
	ActivateWorkspace(ctx context.Context, name string, namespace string) error
	ArchiveWorkspace(ctx context.Context, name string, namespace string) error
	UpdatePhase(ctx context.Context, name string, namespace string, phase string) error
}

// workspaceRepository implements WorkspaceRepository
type workspaceRepository struct {
	NamespacedRepository[*workspacesv1.Workspace, *workspacesv1.WorkspaceList]
	client client.Client
}

// NewWorkspaceRepository creates a new WorkspaceRepository
func NewWorkspaceRepository(k8sClient client.Client) WorkspaceRepository {
	baseRepo := NewK8sNamespacedRepository(
		k8sClient,
		func() *workspacesv1.Workspace { return &workspacesv1.Workspace{} },
		func() *workspacesv1.WorkspaceList { return &workspacesv1.WorkspaceList{} },
	)

	return &workspaceRepository{
		NamespacedRepository: baseRepo,
		client:               k8sClient,
	}
}

// GetByOwner retrieves all workspaces owned by a specific user
func (r *workspaceRepository) GetByOwner(ctx context.Context, owner string, namespace string) (*workspacesv1.WorkspaceList, error) {
	// Use field selector to find workspaces by owner
	return r.List(ctx, namespace, WithFieldSelector("spec.ownedBy="+owner))
}

// GetByWorkMachine retrieves all workspaces in a namespace
// Note: With the 1:1 namespace-to-WorkMachine relationship, this simply returns all workspaces in the namespace
// The workMachineName parameter is kept for backward compatibility but is effectively unused
func (r *workspaceRepository) GetByWorkMachine(ctx context.Context, workMachineName string, namespace string) (*workspacesv1.WorkspaceList, error) {
	// Since there's a 1:1 relationship between namespace and WorkMachine,
	// all workspaces in the namespace belong to the same WorkMachine
	return r.List(ctx, namespace)
}

// ListActive retrieves all active workspaces
func (r *workspaceRepository) ListActive(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error) {
	// Use field selector to find active workspaces
	return r.List(ctx, namespace, WithFieldSelector("spec.status=active"))
}

// ListSuspended retrieves all suspended workspaces
func (r *workspaceRepository) ListSuspended(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error) {
	// Use field selector to find suspended workspaces
	return r.List(ctx, namespace, WithFieldSelector("spec.status=suspended"))
}

// ListArchived retrieves all archived workspaces
func (r *workspaceRepository) ListArchived(ctx context.Context, namespace string) (*workspacesv1.WorkspaceList, error) {
	// Use field selector to find archived workspaces
	return r.List(ctx, namespace, WithFieldSelector("spec.status=archived"))
}

// SuspendWorkspace suspends a workspace by name
func (r *workspaceRepository) SuspendWorkspace(ctx context.Context, name string, namespace string) error {
	// Get the workspace
	workspace, err := r.Get(ctx, namespace, name)
	if err != nil {
		return err
	}

	// Update the status
	workspace.Spec.Status = "suspended"

	// Save the updated workspace
	return r.Update(ctx, workspace)
}

// ActivateWorkspace activates a workspace by name
func (r *workspaceRepository) ActivateWorkspace(ctx context.Context, name string, namespace string) error {
	// Get the workspace
	workspace, err := r.Get(ctx, namespace, name)
	if err != nil {
		return err
	}

	// Update the status
	workspace.Spec.Status = "active"

	// Save the updated workspace
	return r.Update(ctx, workspace)
}

// ArchiveWorkspace archives a workspace by name
func (r *workspaceRepository) ArchiveWorkspace(ctx context.Context, name string, namespace string) error {
	// Get the workspace
	workspace, err := r.Get(ctx, namespace, name)
	if err != nil {
		return err
	}

	// Update the status
	workspace.Spec.Status = "archived"

	// Save the updated workspace
	return r.Update(ctx, workspace)
}

// UpdatePhase updates the phase of a workspace
func (r *workspaceRepository) UpdatePhase(ctx context.Context, name string, namespace string, phase string) error {
	// Get the workspace
	workspace, err := r.Get(ctx, namespace, name)
	if err != nil {
		return err
	}

	// Update the phase in status
	workspace.Status.Phase = phase

	// Use status subresource to update only status
	return r.client.Status().Update(ctx, workspace)
}

File: web/packages/lib/src/k8s/repositories/workspace.repository.ts
import { BaseRepository, type ListOptions } from './base';
import type {
  Workspace,
  WorkspaceList,
  WorkspacePhase,
} from '../types/workspace';
import { buildLabelSelector } from '../utils';
import { parseK8sError } from '../errors';

/**
 * Workspace repository for managing Workspace custom resources
 * Implements all CRUD operations plus workspace-specific operations
 */
export class WorkspaceRepository extends BaseRepository<Workspace> {
  constructor() {
    super('workspaces.kloudlite.io', 'v1', 'workspaces', true);
  }

  /**
   * Get all workspaces owned by a specific user
   */
  async getByOwner(namespace: string, owner: string, options?: ListOptions): Promise<WorkspaceList> {
    const labelSelector = buildLabelSelector({ 'kloudlite.io/owned-by': owner });
    return this.list(namespace, {
      ...options,
      labelSelector,
    }) as Promise<WorkspaceList>;
  }

  /**
   * Get all workspaces in a WorkMachine's namespace
   * Note: With 1:1 namespace-to-WorkMachine relationship, this returns all workspaces in the namespace
   */
  async getByWorkMachine(namespace: string, _workmachineName: string, options?: ListOptions): Promise<WorkspaceList> {
    // Since there's 1:1 relationship between namespace and WorkMachine,
    // all workspaces in the namespace belong to the same WorkMachine
    return this.list(namespace, options) as Promise<WorkspaceList>;
  }

  /**
   * List all workspaces across all namespaces
   */
  async listAll(options?: ListOptions): Promise<WorkspaceList> {
    // Empty namespace means list across all namespaces
    return this.list('', options) as Promise<WorkspaceList>;
  }

  /**
   * List all active workspaces in a namespace
   */
  async listActive(namespace: string, options?: ListOptions): Promise<WorkspaceList> {
    return this.list(namespace, {
      ...options,
      fieldSelector: 'spec.status=active',
    }) as Promise<WorkspaceList>;
  }

  /**
   * List all suspended workspaces in a namespace
   */
  async listSuspended(namespace: string, options?: ListOptions): Promise<WorkspaceList> {
    return this.list(namespace, {
      ...options,
      fieldSelector: 'spec.status=suspended',
    }) as Promise<WorkspaceList>;
  }

  /**
   * List all archived workspaces in a namespace
   */
  async listArchived(namespace: string, options?: ListOptions): Promise<WorkspaceList> {
    return this.list(namespace, {
      ...options,
      fieldSelector: 'spec.status=archived',
    }) as Promise<WorkspaceList>;
  }

  /**
   * Suspend a workspace (stops the workspace pod to save resources)
   */
  async suspend(namespace: string, name: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Update spec.status to suspended
      workspace.spec!.status = 'suspended';

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Activate a workspace (starts the workspace pod)
   */
  async activate(namespace: string, name: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Update spec.status to active
      workspace.spec!.status = 'active';

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Archive a workspace (preserves workspace but makes it inaccessible)
   */
  async archive(namespace: string, name: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Update spec.status to archived
      workspace.spec!.status = 'archived';

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Update workspace phase in status (for controller use)
   */
  async updatePhase(namespace: string, name: string, phase: WorkspacePhase): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Update status.phase
      if (!workspace.status) {
        workspace.status = {};
      }
      workspace.status.phase = phase;

      // Use status subresource to update only status
      return await this.updateStatus(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Connect workspace to an environment
   */
  async connectToEnvironment(
    namespace: string,
    name: string,
    environmentName: string,
    environmentNamespace?: string
  ): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Set environment connection
      workspace.spec!.environmentConnection = {
        environmentRef: {
          kind: 'Environment',
          apiVersion: 'environments.kloudlite.io/v1',
          name: environmentName,
          namespace: environmentNamespace || namespace,
        },
      };

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Disconnect workspace from environment
   */
  async disconnectFromEnvironment(namespace: string, name: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Remove environment connection
      workspace.spec!.environmentConnection = undefined;

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Update workspace settings
   */
  async updateSettings(
    namespace: string,
    name: string,
    settings: Workspace['spec']['settings']
  ): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      // Merge settings
      workspace.spec!.settings = {
        ...workspace.spec!.settings,
        ...settings,
      };

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Get workspace metrics (CPU, memory usage)
   * This is a placeholder - actual implementation would query metrics-server
   */
  async getMetrics(namespace: string, name: string): Promise<{
    cpu: string;
    memory: string;
  } | null> {
    try {
      const workspace = await this.get(namespace, name);

      if (!workspace.status?.resourceUsage) {
        return null;
      }

      return {
        cpu: workspace.status.resourceUsage.cpu || '0',
        memory: workspace.status.resourceUsage.memory || '0',
      };
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Get workspace logs from the workspace pod
   * Returns the pod name - actual log fetching should use Pod API
   */
  async getPodName(namespace: string, name: string): Promise<string | null> {
    try {
      const workspace = await this.get(namespace, name);
      return workspace.status?.podName || null;
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * List workspaces by visibility type
   */
  async listByVisibility(
    namespace: string,
    visibility: 'private' | 'shared' | 'open',
    options?: ListOptions
  ): Promise<WorkspaceList> {
    return this.list(namespace, {
      ...options,
      fieldSelector: `spec.visibility=${visibility}`,
    }) as Promise<WorkspaceList>;
  }

  /**
   * List workspaces shared with a specific user
   */
  async listSharedWith(namespace: string, username: string, options?: ListOptions): Promise<WorkspaceList> {
    // Note: This requires custom filtering as sharedWith is an array
    // For now, we fetch all and filter client-side
    const allWorkspaces = await this.list(namespace, options);

    const filteredItems = allWorkspaces.items.filter(workspace =>
      workspace.spec?.sharedWith?.includes(username)
    );

    return {
      ...allWorkspaces,
      items: filteredItems,
    } as WorkspaceList;
  }

  /**
   * Add user to workspace's sharedWith list
   */
  async shareWith(namespace: string, name: string, username: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      if (!workspace.spec!.sharedWith) {
        workspace.spec!.sharedWith = [];
      }

      if (!workspace.spec!.sharedWith.includes(username)) {
        workspace.spec!.sharedWith.push(username);
      }

      // Update visibility to shared if not already
      if (workspace.spec!.visibility !== 'shared') {
        workspace.spec!.visibility = 'shared';
      }

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }

  /**
   * Remove user from workspace's sharedWith list
   */
  async unshareWith(namespace: string, name: string, username: string): Promise<Workspace> {
    try {
      const workspace = await this.get(namespace, name);

      if (workspace.spec!.sharedWith) {
        workspace.spec!.sharedWith = workspace.spec!.sharedWith.filter(u => u !== username);

        // If no more shared users, change visibility to private
        if (workspace.spec!.sharedWith.length === 0) {
          workspace.spec!.visibility = 'private';
        }
      }

      return await this.update(namespace, name, workspace);
    } catch (err) {
      throw parseK8sError(err);
    }
  }
}

// Export singleton instance
export const workspaceRepository = new WorkspaceRepository();
